### PR TITLE
support PyQt6, remove logic for `compile_qt_svgs`

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
this is going to take some doing (in no small part because of all of our qt-using dependencies)... but this is a start towards supporting PyQt6.

The "hard" bit is done here, by moving away from using compiled resources, and instead using [`QDir.addSearchPath`](https://doc.qt.io/qt-5/qdir.html#addSearchPath).  There's still more to be done and cleaned up; the theme hookspec put resource compiling logic in a number of places, so we need to make sure that's all clean.

the combobox `activated[str]` signal is gone, so a lot of the tests are currently failing due to that.  WIP

### CODE REVIEW

File: napari/utils/theme.py
Comment (tlambert03) [inline napari/utils/theme.py]: 👀 ... this is the key line in this PR
https://doc.qt.io/qt-5/qdir.html#addSearchPath

File: napari/_qt/qt_event_loop.py
Comment (tlambert03) [inline napari/_qt/qt_event_loop.py:200]: 👀 ... this is the key line in this PR
https://doc.qt.io/qt-5/qdir.html#addSearchPath

**Comment by tlambert03** [comment]
OK, I actually think this PR is ready for review.  There's still a bad aesthetic issue with PyQt6 (see image below), but it otherwise generally adds support for pyqt6 and pyside6, so brings a lot to the table.  We don't need to actually add "official" support, but this lets people play with it (cc @psobolewskiPhD).

(can't yet figure out how to fix the background of dock widgets with margins)
Screen Shot 2022-06-11 at 4 54 32 PM

There's a "breaking" change in here in that `compile_qt_svgs` is removed, but I [searched GitHub using codesearch](https://cs.github.com/?scopeName=All+repos&scope=&q=%22compile_qt_svgs%22) and can't find a single example of anyone using it.  I hope we can just drop it without fanfare.

Summary is now:
- adds support for Qt6 by removing internal usage of rcc
- Uses QDir.addSearchPath instead to point to colorized svgs
- removes all of the logic that ran subprocesses to compile qt resources at runtime (yay).
- removes overloaded signals (like`activated[str]` in place of direct connections `currentTextChanged`)
- updates a couple `Qt` namespaces to their nested versions

File: napari/__main__.py
Comment (Czaki) [inline napari/__main__.py]:
```suggestion
    # quick fix for Big Sur py3.9 and qt 5
```

**Comment by tlambert03** [comment]
one more thought: this may negatively impact the bundles, as it tries to write to our own directory at first launch.  the old method did that too i believe, but it's worth checking if anything has changed (permission-wise) here in the bundles.  cc @goanpeca  @jaimergp

File: tox.ini
Comment (Czaki) [inline tox.ini:76]: why this additional deps are required?

**Review by Czaki — APPROVED** [review]
I like this changes. As I see you disable console scripts. This means that console is currently not working with Qt6?

File: tox.ini
Comment (tlambert03) [reply tox.ini:76]: those will only get used when the respective [factors](https://tox.wiki/en/latest/config.html#factors-and-factor-conditional-settings) hit (i.e. when `backend` in the test action is the corresponding value, `pyqt6` or `pyside6`).  The alternative is to put them as extras directly in setup.cfg... but I was leaving that for later, when we actually felt like we wanted to "support" them

**Comment by tlambert03** [comment]
> This means that console is currently not working with Qt6?

oh!  It wasn't last time I checked, but that was months ago, and it appears to be now... Will update

**Comment by psobolewskiPhD** [comment]
Cool! 
In a fresh conda env, if `pyside6` is preinstalled, `pip install -e .` JustWorks on my arm64 mac.
If I tweak setup.cfg then `pip install -e .[pyside6]` can work too!
Same goes for `pyqt6`.

Some minor visual differences—a background perhaps?
image


I'm getting some warnings:
```
WARNING: delivering touch release to same window QWindow(0x0) not QWidgetWindow(0x6000020ac300, name="_QtMainWindowClassWindow")
17:56:17 WARNING delivering touch release to same window QWindow(0x0) not QWidgetWindow(0x6000020ac300, name="_QtMainWindowClassWindow")
WARNING: skipping QEventPoint(id=1 ts=0 pos=0,0 scn=453.284,254.995 gbl=453.284,254.995 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-453.284,-254.995 last=-453.284,-254.995 Δ 453.284,254.995) : no target window
17:56:17 WARNING skipping QEventPoint(id=1 ts=0 pos=0,0 scn=453.284,254.995 gbl=453.284,254.995 Released ellipse=(1x1 ∡ 0) vel=0,0 press=-453.284,-254.995 last=-453.284,-254.995 Δ 453.284,254.995) : no target window
```

(Plus some unrelated warnings from `napari-animation`.)

**Comment by tlambert03** [comment]
> Some minor visual differences—a background perhaps?

yeah, backgrounds of some docked dockwidgets are still expected to look different.  That one’s tricky.

and yes, plugin support will be a per-plugin thing

**Comment by jaimergp** [comment]
> it's worth checking if anything has changed (permission-wise) here in the bundles.

We are defaulting to user directories, so it shouldn't be a problem. If anything, we can precompile the assets as part of installation process (while we can ensure that the user has the right permissions).

**Comment by brisvag** [comment]
Is this still something that needs to run every time, or will it only run on first launch? I'm asking because currently, my [arch-linux napari package](https://aur.archlinux.org/packages/napari) simply raises an error and fails this compilation every time it's launched, the issue being of course permissions to write to the instlalation directory. If this needs to be done only once, then it would be nice to use this machinery on installation -- rather than on launch -- when we have permissions.

**Comment by tlambert03** [comment]
this one has been ready to go for about a month.  We have one approval, but would would be nice to get another.  I'm not sure when to merge this one, but will probably do so in about a week.  @napari/core-devs, if anyone is interested, please have a look before then

**Review by sofroniewn — APPROVED** [review]
Fine with me - nice work here @tlambert03. Looks good to merge when you're ready

File: napari/_qt/_tests/test_qt_viewer.py
Comment (andy-sweet) [inline napari/_qt/_tests/test_qt_viewer.py]: Is this related to Qt (6)? Or is this an extra change you slipped into this PR?

File: napari/_qt/layer_controls/qt_points_controls.py
Comment (andy-sweet) [inline napari/_qt/layer_controls/qt_points_controls.py]: Is this change required for Qt6?

Optional: I personally prefer the explicit comparison rather the truthy cast that would also make `Qt.PartiallyChecked => True` even though we shouldn't ever get that value right now.

File: napari/_qt/layer_controls/qt_vectors_controls.py
Comment (andy-sweet) [inline napari/_qt/layer_controls/qt_vectors_controls.py:118]: Was the position of this actually causing issues? If so, you should add an explanatory implementation comment.

Or do you just prefer these layer event connections to go last? If so, no need to do anything.

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py]:
```suggestion
    def event(self, e: QEvent) -> bool:
```

Kinda obvious, but since you're updating the return type and we imported `QEvent`, might as well?

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py:153]: Does this change catch both the name/type change from Qt5 to Qt6, but also because only some events have positions (e.g. `QMouseEvent`)? Is `QMainWindow.event` guaranteed to only get `QMouseEvents`? Regardless, a short explanatory comment here would be useful to explain why we need this check and that we could remove it if we ever deprecated Qt5 (unlikely...).

File: napari/_qt/qt_resources/styles/02_custom.qss
Comment (andy-sweet) [inline napari/_qt/qt_resources/styles/02_custom.qss:39]: Is this relevant to Qt5->6? Or just a bug fix?

File: napari/_qt/qt_viewer.py
Comment (andy-sweet) [inline napari/_qt/qt_viewer.py:729]: Is there a reason you removed the keywords for the arguments? I can usually stomach 1 and sometimes 2 (depending on the function) positional arguments from a readability standpoint.

File: napari/_qt/widgets/qt_action_context_menu.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_action_context_menu.py:110]: Is this an independent bug fix?

File: napari/_qt/widgets/qt_scrollbar.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_scrollbar.py:26]: See similar comment in `_qt_main_window.py` - I can't link to it before submitting the review.

File: napari/resources/_icons.py
Comment (andy-sweet) [inline napari/resources/_icons.py]:
```suggestion
    combination (Cartesian product) of `svg_path`, `color`, and `opacity`
```

More respect for M. Descartes please.

File: napari/resources/_icons.py
Comment (andy-sweet) [inline napari/resources/_icons.py:109]:
```suggestion
        Resource system (such as qss), and `xml` is the *raw* colorzied SVG
```

You seem to mostly be using lowercase for acronyms, so prob best to stick with it.

**Review by andy-sweet — APPROVED** [review]
Looks good to me. I just had a bunch of annoying clarifying questions and nitpicks.

**Comment by andy-sweet** [comment]
Also, to clarify this only adds support for Qt6, but we're not updating any of the setup files to depend on it yet right? I guess we'll wait for broader availability on conda-forge/PyPI before doing that?

**Comment by Czaki** [comment]
For PyPi (setup.cfg) we do not need to update anything.

File: napari/_qt/_tests/test_qt_viewer.py
Comment (tlambert03) [reply napari/_qt/_tests/test_qt_viewer.py]: good catch, that looks like a git merge mistake (this is an old PR that's gone through many merges).  I'll fix this and double check for other possible merge mistakes

File: napari/_qt/widgets/qt_action_context_menu.py
Comment (tlambert03) [reply napari/_qt/widgets/qt_action_context_menu.py:110]: nope, `menu()` is gone in qt6, and it turns out it's not necessary here.

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py:153]: > Is QMainWindow.event guaranteed to only get QMouseEvents?

no, but if the type() is a `QEvent.Type.ToolTip` type as checked here, then it will have position info

File: napari/_qt/widgets/qt_scrollbar.py
Comment (tlambert03) [reply napari/_qt/widgets/qt_scrollbar.py:26]: added comment and link to qt deprecation notice

File: napari/_qt/layer_controls/qt_points_controls.py
Comment (tlambert03) [reply napari/_qt/layer_controls/qt_points_controls.py]: > the truthy cast that would also make Qt.PartiallyChecked => True

yeah, I think that's what was actually necessary here.  If I revert this change, I get a test fail on Qt6.  I've changed it to `state != Qt.Unchecked`

File: napari/_qt/qt_resources/styles/02_custom.qss
Comment (tlambert03) [reply napari/_qt/qt_resources/styles/02_custom.qss:39]: that was a relevant change for styles in Qt6

File: napari/_qt/qt_viewer.py
Comment (tlambert03) [reply napari/_qt/qt_viewer.py:729]: yeah: unfortunately, PySide and PyQt5 don't always wrap these things in exactly the same way. so there are cases where using keyword instead of positional arguments will throw an error in one of the backends.   I can't remember now if I found this due to that ... presumably i did.  but either way, this is unfortunately the safer route to take with pyside/pyqt (though I'm with you on the readability)

File: napari/_qt/layer_controls/qt_vectors_controls.py
Comment (tlambert03) [reply napari/_qt/layer_controls/qt_vectors_controls.py:118]: it was actually causing issues (the way I did this PR was just to use pyqt6/pyside6 and keep running tests until it passed) by triggering an event too soon. 
Presumably this was due to some change in the order of event connections. I will add a comment, but it also could have just as easily been written differently to begin with (as it is in many other places) and we'd never have noticed (it's a bit of an arbitrary place to put a comment about it).  It will show up in tests though.

**Review by tlambert03 — COMMENTED** [review]
thanks as always for the thorough review @andy-sweet

pretty much everything you noticed here did indeed need to be changed for pyqt6.  I've added a couple more comments for clarification

**Comment by tlambert03** [comment]
> Also, to clarify this only adds support for Qt6, but we're not updating any of the setup files to depend on it yet right? I guess we'll wait for broader availability on conda-forge/PyPI before doing that?

correct.  and I would say it adds "experimental" support for Qt6.  As shown above, there's still a cosmetic issue with Qt6 that I can't figure out (but which doesn't break any functionality).  At minimum, that would need to be fixed before we call Qt6 fully supported

File: napari/_qt/layer_controls/qt_points_controls.py
Comment (tlambert03) [reply napari/_qt/layer_controls/qt_points_controls.py]: actually, looks like even `state != Qt.Unchecked` doesn't work.  it _must_ be cast to bool to work on Qt6.   I suppose they've overloaded the `==` operator to return something that is not a bool in qt6?

Comment (andy-sweet) [reply napari/_qt/layer_controls/qt_points_controls.py]: My only other thought is that Qt6 might require `state != Qt.CheckState.Unchecked`, but I don't have any evidence for that. Using the truthy cast is fine.

File: napari/_qt/layer_controls/qt_vectors_controls.py
Comment (andy-sweet) [reply napari/_qt/layer_controls/qt_vectors_controls.py:118]: If there is a failing test, no need to add the explanatory comment.

**Comment by andy-sweet** [comment]
Looks good to me. Feel free to merge when you want. Probably soonish to avoid conflicts.

File: napari/_qt/layer_controls/qt_points_controls.py
Comment (tlambert03) [reply napari/_qt/layer_controls/qt_points_controls.py]: figured it out.  On PyQt6, (but not pyside6) Qt.CheckState members do not equate to their integer equivalents, morever you can't even _cast_ them to an int, but rather have to use `enum.value`, which isn't the case for PyQt5 (I remember reading somewhere that PyQt6 enums are now proper python enums, but unfortunately they apparently didn't use an `IntEnum` :joy:

```python
>>> from PyQt6.QtCore import Qt
>>> Qt.CheckState.Unchecked
<CheckState.Unchecked: 0>
>>> int(Qt.CheckState.Unchecked)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'CheckState'
>>> 0 == Qt.CheckState.Unchecked
False  # annoying
>>> 0 == Qt.CheckState.Unchecked.value
True
```

```python
>>> from PySide6.QtCore import Qt
>>> int(Qt.CheckState.Unchecked)
0
>>> 0 == Qt.CheckState.Unchecked
True
```

so, summary is that we could either case the `state` integer into the corresponding enum type for whatever backend we're using (may be a Qt enum or a python enum, depends), or we can simply just go with `bool(state)` ... 
so I'm going with it :)

**Pull request merged**
